File: backend/src/main/java/coursepick/coursepick/domain/Distance.java
package coursepick.coursepick.domain;

public record Distance(
        double meter
) {
    public static Distance zero() {
        return new Distance(0);
    }

    public static Distance max() {
        return new Distance(Double.MAX_VALUE);
    }

    public Distance add(Distance other) {
        return new Distance(this.meter() + other.meter());
    }

    public Distance minimum(Distance other) {
        return this.meter() <= other.meter() ? this : other;
    }
}

File: android/app/src/main/java/io/coursepick/coursepick/domain/course/Distance.kt
package io.coursepick.coursepick.domain.course

@JvmInline
value class Distance(
    val meter: Meter,
) : Comparable<Distance> {
    init {
        require(meter >= 0)
    }

    constructor(meter: Double) : this(Meter(meter))

    override fun compareTo(other: Distance): Int = this.meter.compareTo(other.meter)

    companion object {
        operator fun invoke(meter: Int): Distance = Distance(Meter(meter))
    }
}
